File: iot-device/src/crypto/message_crypto.cpp
/**
 * Message Encryption Module
 * WARNING: This code contains intentional security vulnerabilities for testing purposes
 * DO NOT use in production
 */

#include <openssl/evp.h>
#include <openssl/rand.h>
#include <cstring>
#include <vector>

/**
 * Simple XOR cipher for message encryption
 * VULNERABLE: Custom XOR encryption (insecure)
 */
void xor_encrypt(const unsigned char* plaintext, size_t length,
                 const unsigned char* key, size_t key_length,
                 unsigned char* ciphertext) {
    // VULNERABLE: XOR cipher is not secure
    for (size_t i = 0; i < length; i++) {
        ciphertext[i] = plaintext[i] ^ key[i % key_length];
    }
}

/**
 * XOR decryption (same as encryption)
 * VULNERABLE: XOR cipher
 */
void xor_decrypt(const unsigned char* ciphertext, size_t length,
                 const unsigned char* key, size_t key_length,
                 unsigned char* plaintext) {
    // VULNERABLE: XOR cipher
    xor_encrypt(ciphertext, length, key, key_length, plaintext);
}

/**
 * Simple substitution cipher
 * VULNERABLE: Custom weak cipher
 */
void substitute_encrypt(unsigned char* data, size_t length) {
    // VULNERABLE: Simple substitution cipher (easily broken)
    for (size_t i = 0; i < length; i++) {
        data[i] = (data[i] + 13) % 256; // ROT13-like for bytes
    }
}

/**
 * Substitute decrypt
 * VULNERABLE: Reverse of substitution
 */
void substitute_decrypt(unsigned char* data, size_t length) {
    // VULNERABLE: Simple substitution reversal
    for (size_t i = 0; i < length; i++) {
        data[i] = (data[i] - 13 + 256) % 256;
    }
}

/**
 * Weak stream cipher
 * VULNERABLE: Custom stream cipher (not secure)
 */
class WeakStreamCipher {
private:
    unsigned char state[256];
    int i, j;

public:
    // VULNERABLE: Custom stream cipher implementation
    WeakStreamCipher(const unsigned char* key, size_t key_length) {
        // Initialize state
        for (int k = 0; k < 256; k++) {
            state[k] = k;
        }

        // Key scheduling (similar to RC4 but simplified)
        j = 0;
        for (int k = 0; k < 256; k++) {
            j = (j + state[k] + key[k % key_length]) % 256;
            std::swap(state[k], state[j]);
        }

        i = 0;
        j = 0;
    }

    unsigned char get_byte() {
        // VULNERABLE: Weak pseudo-random byte generation
        i = (i + 1) % 256;
        j = (j + state[i]) % 256;
        std::swap(state[i], state[j]);
        return state[(state[i] + state[j]) % 256];
    }

    void encrypt(unsigned char* data, size_t length) {
        // VULNERABLE: Stream cipher encryption
        for (size_t k = 0; k < length; k++) {
            data[k] ^= get_byte();
        }
    }
};

/**
 * Encrypt message with weak cipher
 * VULNERABLE: Using custom weak stream cipher
 */
void encrypt_message(const unsigned char* plaintext, size_t length,
                    const unsigned char* key, size_t key_length,
                    unsigned char* ciphertext) {
    // VULNERABLE: Custom stream cipher
    memcpy(ciphertext, plaintext, length);
    WeakStreamCipher cipher(key, key_length);
    cipher.encrypt(ciphertext, length);
}

/**
 * Decrypt message with weak cipher
 * VULNERABLE: Decryption using weak cipher
 */
void decrypt_message(const unsigned char* ciphertext, size_t length,
                    const unsigned char* key, size_t key_length,
                    unsigned char* plaintext) {
    // VULNERABLE: Stream cipher decryption (same as encryption)
    memcpy(plaintext, ciphertext, length);
    WeakStreamCipher cipher(key, key_length);
    cipher.encrypt(plaintext, length);
}

/**
 * Encrypt sensor data with XOR
 * VULNERABLE: XOR encryption for sensor data
 */
void encrypt_sensor_data(const unsigned char* data, size_t length,
                        unsigned char* encrypted) {
    // VULNERABLE: Fixed XOR key
    const unsigned char key[] = "sensor_key_12345";
    xor_encrypt(data, length, key, sizeof(key) - 1, encrypted);
}

/**
 * Obfuscate configuration
 * VULNERABLE: Simple obfuscation (not encryption)
 */
void obfuscate_config(unsigned char* config, size_t length) {
    // VULNERABLE: Simple bit manipulation (not secure)
    for (size_t i = 0; i < length; i++) {
        config[i] = ~config[i]; // Just flip bits
    }
}

File: crypto-messenger/app/crypto/message_crypto.py
"""
Message Encryption Module
WARNING: This code contains intentional security vulnerabilities for testing purposes
DO NOT use in production
"""

from Crypto.Cipher import ARC4
import hashlib
import os


class MessageCrypto:
    """Vulnerable message encryption using RC4"""

    def __init__(self, key=None):
        # VULNERABLE: Using RC4 cipher (broken)
        self.key = key or b'default_weak_key_12345'

    def encrypt_message(self, plaintext):
        """
        Encrypt message using RC4 stream cipher
        VULNERABLE: RC4 is broken and should not be used
        """
        # VULNERABLE: RC4 cipher usage
        cipher = ARC4.new(self.key)
        ciphertext = cipher.encrypt(plaintext.encode())
        return ciphertext.hex()

    def decrypt_message(self, ciphertext_hex):
        """
        Decrypt message using RC4
        VULNERABLE: RC4 decryption
        """
        # VULNERABLE: RC4 cipher usage
        ciphertext = bytes.fromhex(ciphertext_hex)
        cipher = ARC4.new(self.key)
        plaintext = cipher.decrypt(ciphertext)
        return plaintext.decode()

    def encrypt_file(self, file_data):
        """
        Encrypt file data using RC4
        VULNERABLE: RC4 for file encryption
        """
        # VULNERABLE: RC4 stream cipher
        cipher = ARC4.new(self.key)
        encrypted = cipher.encrypt(file_data)
        return encrypted


class CustomCrypto:
    """Custom encryption implementation (insecure)"""

    @staticmethod
    def xor_encrypt(data, key):
        """
        Custom XOR encryption
        VULNERABLE: Custom crypto, XOR cipher is weak
        """
        # VULNERABLE: Custom encryption algorithm (bad practice)
        # VULNERABLE: Simple XOR is not secure
        key_bytes = key.encode() if isinstance(key, str) else key
        data_bytes = data.encode() if isinstance(data, str) else data

        encrypted = bytearray()
        for i, byte in enumerate(data_bytes):
            encrypted.append(byte ^ key_bytes[i % len(key_bytes)])

        return bytes(encrypted).hex()

    @staticmethod
    def xor_decrypt(encrypted_hex, key):
        """
        Custom XOR decryption
        VULNERABLE: Symmetric with encryption, weak algorithm
        """
        # VULNERABLE: XOR decryption (same as encryption)
        encrypted = bytes.fromhex(encrypted_hex)
        return CustomCrypto.xor_encrypt(encrypted, key)

    @staticmethod
    def simple_hash(data):
        """
        Simple custom hash function
        VULNERABLE: Custom hash implementation
        """
        # VULNERABLE: Custom hash function (never do this)
        result = 0
        for char in data:
            result = (result * 31 + ord(char)) % (2**32)
        return hex(result)[2:]


class HardcodedKeys:
    """Class with hardcoded cryptographic keys"""

    # VULNERABLE: Hardcoded encryption keys
    DEFAULT_KEY = b'supersecretkey12345'
    MASTER_KEY = b'master_encryption_key_hardcoded'
    API_SECRET = 'api_secret_12345_hardcoded'

    # VULNERABLE: Hardcoded salt for key derivation
    SALT = b'fixed_salt_value'

    @staticmethod
    def get_encryption_key():
        """
        Return hardcoded encryption key
        VULNERABLE: Hardcoded key in source code
        """
        # VULNERABLE: Returning hardcoded key
        return HardcodedKeys.DEFAULT_KEY

    @staticmethod
    def derive_key(password):
        """
        Derive key from password with hardcoded salt
        VULNERABLE: Hardcoded salt makes this predictable
        """
        # VULNERABLE: Using hardcoded salt
        # VULNERABLE: Using MD5 for key derivation
        key = hashlib.md5(password.encode() + HardcodedKeys.SALT).digest()
        return key
